tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -741,6 +755,691 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
+
+	/*
+	 * Set the relation index.  This is defined as the position of this
+	 * joinrel in the join_rel_list list plus the length of the rtable list.
+	 * Note that since this joinrel is at the end of the join_rel_list list
+	 * when we are called, we can get the position by list_length.
+	 */
+	fpinfo->relation_index = list_length(root->parse->rtable) + list_length(root->join_rel_list);
+
+	return true;
+}
+
+#ifdef USE_ASSERT_CHECKING
+/*
+ * Check if the given hypertable is a distributed hypertable.
+ */
+static bool
+is_distributed_hypertable(Oid hypertable_id)
+{
+	Cache *hcache;
+
+	Hypertable *ht =
+		ts_hypertable_cache_get_cache_and_entry(hypertable_id, CACHE_FLAG_MISSING_OK, &hcache);
+
+	/* perform check before cache is released */
+	bool ht_is_distributed = (ht != NULL && hypertable_is_distributed(ht));
+	ts_cache_release(hcache);
+
+	return ht_is_distributed;
+}
+#endif
+
+/*
+ * Create a new join partition RelOptInfo data structure for a partition. The data
+ * structure is based on the parameter joinrel. The paramter is taken as template
+ * and adjusted for the partition provided by the parameter hyper_table_rel.
+ */
+static RelOptInfo *
+create_join_partition(PlannerInfo *root, RelOptInfo *innerrel, RelOptInfo *joinrel,
+					  RelOptInfo *hyper_table_rel, AppendRelInfo *appinfo)
+{
+	RelOptInfo *join_partition = palloc(sizeof(RelOptInfo));
+	memcpy(join_partition, joinrel, sizeof(RelOptInfo));
+
+	/* Create a new relinfo for the join partition */
+	join_partition->fdw_private = NULL;
+	TsFdwRelInfo *join_part_fpinfo = fdw_relinfo_create(root,
+														join_partition,
+														hyper_table_rel->serverid,
+														InvalidOid,
+														TS_FDW_RELINFO_REFJOIN_PARTITION);
+
+	Assert(join_part_fpinfo != NULL);
+
+	TsFdwRelInfo *hyper_table_rel_fpinfo = fdw_relinfo_get(hyper_table_rel);
+	Assert(hyper_table_rel_fpinfo != NULL);
+
+	/* Copy chunk assignment from hypertable */
+	join_part_fpinfo->sca = hyper_table_rel_fpinfo->sca;
+
+	/* Set parameters of the join partition */
+	join_partition->relid = hyper_table_rel->relid;
+	join_partition->relids = bms_copy(hyper_table_rel->relids);
+	join_partition->relids = bms_add_members(join_partition->relids, innerrel->relids);
+	join_partition->pathlist = NIL;
+	join_partition->partial_pathlist = NIL;
+
+	/* Set the reltarget expressions of the partition based on the reltarget expressions
+	 * of the join and adjust them for the partition */
+	join_partition->reltarget = create_empty_pathtarget();
+	join_partition->reltarget->sortgrouprefs = joinrel->reltarget->sortgrouprefs;
+	join_partition->reltarget->cost = joinrel->reltarget->cost;
+	join_partition->reltarget->width = joinrel->reltarget->width;
+#if PG14_GE
+	join_partition->reltarget->has_volatile_expr = joinrel->reltarget->has_volatile_expr;
+#endif
+	join_partition->reltarget->exprs =
+		castNode(List,
+				 adjust_appendrel_attrs(root, (Node *) joinrel->reltarget->exprs, 1, &appinfo));
```

Comment:
Yes, this is needed. The join is performed partition-wise. Each of the partitions belongs to one data node and we are adjusting the expressions here using the `appinfo` of the data node.

Reply:
Let's call it "data-node-wise" or something, because "partition" is ambiguous between chunks and data nodes. So which kind of adjustment exactly do we do? I see that `adjust_apprendrel_attrs` can change varno or convert row type. For data node joinrels, the varno and rowtype are the same, because we just copy them from the base joinrel. Does this function do anything else?

Reply:
The function is renamed as suggested.

With `adjust_apprendrel_attrs`, the relid in the expressions is translated from the hypertable relid to the relid of the data node that is responsible for this partition. Otherwise, the deparser could not pushdown these expressions properly.

Reply:
> With adjust_apprendrel_attrs, the relid in the expressions is translated from the hypertable relid to the relid of the data node that is responsible for this partition.

Yes, but aren't they exactly equal? As set here: https://github.com/timescale/timescaledb/pull/5212/files#diff-47f57773f9558fb5ef0345e680eb19c0d1c9378c1701011b7a930a96e8a0813cR1135

This is relid in the sense of `varno` and `RelOptInfo`, not the oid from RTE.

Reply:
Yes, these values are equal. However, in this function, we have to adjust the expressions for the join partition and make sure that the expressions (they are generated for the unpartitioned join) belong to this partition. We do something similar in [adjust_data_node_rel_attrs](https://github.com/timescale/timescaledb/blob/46632f887365a3c64377571741114fc4cd11a7d6/tsl/src/fdw/data_node_scan_plan.c#L217) for the DataNodes.

Reply:
OK, I still don't understand what exactly it adjusts, given that the varno and row type are exactly the same, but I'll have to trust you on this :)

Reply:
Sorry for the imprecise answer. The `varno` is adjusted at this point 

For example, here you see the `exprs` before and after the adjustment. The `varno` of the expressions is translated from the input table position to the proper position of this specific DataNode entry in the simple_rel_array (the data node is at position 8 in this example). This is needed for the deparser to generate the SQL for the partition properly.

### Original List
```
(
   {VAR 
   :varno 1 
   :varattno 2 
   :vartype 23 
   :vartypmod -1 
   :varcollid 0 
   :varlevelsup 0 
   :varnosyn 1 
   :varattnosyn 2 
   :location -1
   [...]
)
```

### Adjusted List
```
(
   {VAR 
   :varno 8 
   :varattno 2 
   :vartype 23 
   :vartypmod -1 
   :varcollid 0 
   :varlevelsup 0 
   :varnosyn 8 
   :varattnosyn 2 
   :location -1
   }
   [...]
)
```